frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx

```diff
@@ -69,10 +84,100 @@ import "./NativeQueryEditor.css";
 const AUTOCOMPLETE_DEBOUNCE_DURATION = 700;
 const AUTOCOMPLETE_CACHE_DURATION = AUTOCOMPLETE_DEBOUNCE_DURATION * 1.2; // tolerate 20%
 
-export class NativeQueryEditor extends Component {
+type SidebarFeatures = {
+  dataReference?: boolean;
+  variables?: boolean;
+  snippets?: boolean;
+  promptInput?: boolean;
+};
+
+type CardCompletionItem = Pick<Card, "id" | "name" | "dataset"> & {
+  collection_name: string;
+};
+
+type AutocompleteItem = [string, string];
+
+interface OwnProps {
+  question: Question;
+  query: NativeQuery;
+
+  nativeEditorSelectedText?: string;
+
+  snippets?: NativeQuerySnippet[];
+  modalSnippet?: NativeQuerySnippet;
+  snippetCollections?: Collection[];
+
+  viewHeight: number;
+  width: number;
+
+  isOpen?: boolean;
+  isInitiallyOpen?: boolean;
+  isNativeEditorOpen: boolean;
+  isRunning?: boolean;
+
+  readOnly?: boolean;
+  enableRun?: boolean;
+  canChangeDatabase?: boolean;
+  canUsePromptInput?: boolean;
+  cancelQueryOnLeave?: boolean;
+  hasTopBar?: boolean;
+  hasParametersList?: boolean;
+  hasEditingSidebar?: boolean;
+  sidebarFeatures?: SidebarFeatures;
+  resizable?: boolean;
+  resizableBoxProps?: Partial<ResizableBoxProps>;
+
+  editorContext?: "question";
+
+  handleResize: () => void;
+  autocompleteResultsFn: (prefix: string) => Promise<AutocompleteItem[]>;
+  cardAutocompleteResultsFn: (prefix: string) => Promise<CardCompletionItem[]>;
+  setDatasetQuery: (query: NativeQuery) => Promise<Question>;
+  fetchQuestion: (cardId: CardId) => Promise<Card>;
+  runQuestionQuery: (opts?: {
+    overrideWithCard?: Card;
+    shouldUpdateUrl?: boolean;
+  }) => void;
+  setNativeEditorSelectedRange: (range: Ace.Range) => void;
+  openDataReferenceAtQuestion: (id: CardId) => void;
+  openSnippetModalWithSelectedText?: (text: string) => void;
+  insertSnippet: (snippet: NativeQuerySnippet) => void;
+  setIsNativeEditorOpen?: (isOpen: boolean) => void;
+  setParameterValue: (parameterId: ParameterId, value: string) => void;
+  cancelQuery?: () => void;
+  closeSnippetModal: () => void;
+}
+
+type Props = OwnProps;
+
+interface NativeQueryEditorState {
+  initialHeight: number;
+  isSelectedTextPopoverOpen: boolean;
+  mobileShowParameterList: boolean;
+  isPromptInputVisible: boolean;
+}
+
+type AceCompletionsGetter = Ace.Completer["getCompletions"];
+
+export class NativeQueryEditor extends Component<
+  Props,
+  NativeQueryEditorState
+> {
+  editor = createRef<HTMLDivElement>();
+  resizeBox = createRef<HTMLDivElement & ResizableBox>();
+
+  // this is overwritten when the editor mounts
+  nextCompleters?: (position: Ace.Position) => Ace.Completer[] = undefined;
+
+  _editor: Ace.Editor | null = null;
+  _lastAutoComplete: {
+    timestamp?: number;
+    prefix?: string | null;
+    results?: AutocompleteItem[];
+  } = {};
```

Comment:
This is more accurate type definition for it:

```suggestion
  _lastAutoComplete?: {
    timestamp: number;
    prefix: string | null;
    results: AutocompleteItem[];
  };
```


But I think it's safe and better to use a default value here (the same as in `loadAceEditor()`):

```suggestion
  _lastAutoComplete: {
    timestamp: number;
    prefix: string | null;
    results: AutocompleteItem[];
  } = { timestamp: 0, prefix: null, results: [] };
```